sub/sd_ass.c

```diff
@@ -756,65 +748,86 @@ static void ass_to_plaintext(struct buf *b, const char *in)
         }
     }
     // A '{' without a closing '}' is always visible.
-    if (in_tag) {
-        while (*open_tag_pos)
-            append(b, *open_tag_pos++);
+    if (open_tag_pos) {
+        bstr_xappend(NULL, b, bstr0(open_tag_pos));
     }
 }
 
-// Empty string counts as whitespace. Reads s[len-1] even if there are \0s.
-static bool is_whitespace_only(char *s, int len)
+// Empty string counts as whitespace.
+static bool is_whitespace_only(bstr b)
 {
-    for (int n = 0; n < len; n++) {
-        if (s[n] != ' ' && s[n] != '\t')
+    for (int n = 0; n < b.len; n++) {
+        if (b.start[n] != ' ' && b.start[n] != '\t')
             return false;
     }
     return true;
 }
 
-static char *get_text_buf(struct sd *sd, double pts, enum sd_text_type type)
+static bstr get_text_buf(struct sd *sd, double pts, enum sd_text_type type)
 {
     struct sd_ass_priv *ctx = sd->priv;
     ASS_Track *track = ctx->ass_track;
 
     if (pts == MP_NOPTS_VALUE)
-        return NULL;
+        return (bstr){0};
     long long ipts = find_timestamp(sd, pts);
 
-    struct buf b = {ctx->last_text, sizeof(ctx->last_text) - 1};
+    bstr *b = &ctx->last_text;
+
+    if (!b->start)
+        b->start = talloc_size(ctx, 4096);
```

Comment:
No need to preallocate anything, it will handle itself fine.

Reply:
This is deliberate; explicitly allocating here means that `last_text.start` will have `ctx` as its parent (without needing to pass it into every function that interacts with `b` later, including in eg `ass_to_plaintext`), which ensures it'll be freed when the sd is torn down.

Reply:
But it doesn't ensure that it won't be reallocated if the size become > 4096

Reply:
`ta_realloc_size` preserves the existing allocation's parent when the existing pointer is non-null.

Reply:
of course if the size is enough, but it seems fragile in case the size happens to be bigger at some point.

Reply:
No, like, if `ta_realloc_size` is called to expand `last_text->start` to a larger size, the new larger allocation will keep the existing parent.

Reply:
ah, you meant parent, I can't read. Fair.